### ISSUE TITLE
HTTTP/2 flood prevention

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#1346]
## Scope

[Low-Rate Denial-of-Service Attacks against HTTP/2 Services](https://www.researchgate.net/publication/280591671_Low-Rate_Denial-of-Service_Attacks_against_HTTP2_Services) paper studies HTTP/2 floods using `PING` and `WINDOW_UPDATE` frames. It's not good to fight against the attacks using packet rate limit since a client can send large POST requests for example. From the other hand requests rate in such attacks could be zero. So a new limits *with burst*, just the same as for current request limiting, must be developed.

## Testing

The test must be developed as part of #673 .

## Documentation

Please update https://github.com/tempesta-tech/tempesta/wiki/HTTP-security#frang-security-limits-enforcing-module with the reference to the paper.

**Comment by RomanBelozerov** [comment]
> The HTTP limits must be extended by stream_rate, stream_burst and concurrent_streams just similar to equal connection limits.

Tempesta should send SETTINGS_MAX_CONCURRENT_STREAMS in SETTINGS frame when initializing connection with client.

**Comment by EvgeniiMekhanik** [comment]
During developing this task we should don't forget to change
```C
if (closed_streams->num <= TFW_MAX_CLOSED_STREAMS
    || (max_streams == ctx->streams_num
	&& closed_streams->num))
```
in `tfw_h2_closed_streams_shrink()`

By @EvgeniiMekhanik  : It was already fixed long time ago. https://github.com/tempesta-tech/tempesta/commit/14cd37b67932b9fe248bea98666be9bc6f4e5b81. Moreover this comment is incorrect. During fixing this problem it was found that we should use another condition.

**Comment by krizhanovsky** [comment]
Related CVE and security advisory https://github.com/advisories/GHSA-93p3-5r25-4p75

**Comment by const-t** [comment]
1. First of all we must not implement `stream_rate`, `stream_burst`, we have `request_rate/burst` that works the same.
2. `response_timeout` it's very doubtful. It can help us only if limit whole connection lifetime, without any updates during connection lifetime. For example: Slow read client opens connection with 100 streams, stream 1 sends request, Tempesta scheduled it to upstream. Start `response_timeout`, if response timeout exceeded, reset connection.
Cons of this method:   
    1. We always must know how many time will take transferring of largest file with slowest client.
    2. It's per-connection not per-stream.

However, it seems in current task we must implement rate limits for SETTINGS, PING and maybe WINDOW_UPDATE frame. But let's start from SETTINGS and PING. WINDOW_UPDATE requires some discussion, because it's not so easy and maybe it not worth it. For instance, client can send unlimited number of WINDOW_UPDATE frames to closed streams, that also not good.

**Comment by const-t** [comment]
After some discussion, we came to the conclusion that we need one timer per connection and bitfield that indicates stream stream being read in the timer interval. On every shot timer clears bitfield. However this doesn't look like complete solution.

**Comment by kingluo** [comment]
For "HTTTP/2 slow read":

The slow-read attack affects the CPU, memory, and incoming connection pool. So #498 does not cover the issue.

A simple solution is:

1. define two thresholds: `slow_read_win_lower_limit`, like 10 bytes, and another threshold `slow_read_response_timeout`, like 10 secs.
2. if the HTTP/2 receive window (resulting from `WINDOW_UPDATE`) of the client is lower than `slow_read_win_lower_limit`, we count the window duration time.
3. if the accumulated time reaches `slow_read_response_timeout`, we abort the TCP connection.
4. these procedures are also suitable for request direction, then `response_slow_read_time` becomes `request_slow_read_time`.
5. the time measurement is per stream, but the timer needed by step 3 is per connection.
6. optional: maybe we offload timeout action to Linux worker queue to eliminate the side effect of timers.

**Comment by kingluo** [comment]
**For "HTTTP/2 flood prevention":**
I doubt the low-rate service frames like `PING` can form the `DOS`, so maybe we just focus on the high-rate case? That is the `PING` flood, as well as `SETTING` (as suggested by @const-t, we can postpone handling `WINDOW_UPDATE`).

**A simple solution:**
Maybe we just maintain the accounting about `PING` and `SETTING` just like `HEADERS` and check it in frang. I need to investigate the code more to make it detailed.

**Comment by const-t** [comment]
In this issue being implemented must fix following CVEs:
- [ ] CVE-2019-9511 “Data Dribble” (https://github.com/tempesta-tech/tempesta-test/pull/618)
- [ ] CVE-2019-9517 “Internal Data Buffering” (https://github.com/tempesta-tech/tempesta-test/pull/618)
- [x] CVE-2019-9512 “Ping Flood” (https://github.com/tempesta-tech/tempesta-test/pull/729)
- [ ] CVE-2019-9515 “Settings Flood” (https://github.com/tempesta-tech/tempesta-test/pull/618/changes#diff-5748e9c9bb08b07ce68e6dd0b9c64f17ce4ad60533612cd0e3dbcb01e9122d30R310)

- [ ] @krizhanovsky : also this change shall protect against [CDN amplification attack](https://www.usenix.org/conference/usenixsecurity24/presentation/lin-ziyu). For now we have issue in webshield https://github.com/tempesta-tech/escudo/issues/441 , but the logic seems heavyweight and it'd be better to protect on Tempesta FW side.

**Comment by kingluo** [comment]
@krizhanovsky @const-t @RomanBelozerov 
The exploit of [CVE-2019-9512](https://github.com/advisories/GHSA-hgr8-6h9x-f7q9), [CVE-2019-9515](https://github.com/advisories/GHSA-9259-5376-vjcj) and [CVE-2019-9514](https://github.com/advisories/GHSA-39qc-96h7-956f) has a common precondition: the implementation has a queue to accumulate the control frames.

> Some HTTP/2 implementations are vulnerable to ping floods, potentially leading to a denial of service. The attacker sends continual pings to an HTTP/2 peer, causing the peer to build an internal queue of responses. Depending on how efficiently this data is queued, this can consume excess CPU, memory, or both.

However, tempesta does not, it just sends or replies to the frames directly into the TCP write queue of the socket, and that queue is throttled by the TCP flow control, so I don't think there is any problem there.

I recommend writing test cases first for these CVEs and verifying if there is a problem, if not, then we just ignore them.

**Comment by krizhanovsky** [comment]
Crucial since DDoS affects our installation

**Comment by krizhanovsky** [comment]
Minutes on today call:
* there are different cases for control frames DDoS (e.g. PING): many skbs per a frame and we fail on memory and many frames per skb and we fail on CPU. The second case can not be handled with `client_mem`
* thus, we still need rate limit on control frames
* sending 0 TCP window is dangerous, which can lead to connection deadlock. In #498 we're going to control flow and backpreasure with TCP send/receive buffers (not evict skb from the queues)
* #1346 should be fixes as slow read with a timer and control frames wit ingress rate limit
* #2456 is for `client_mem` `hard_limit` still worth having for the new mem exhausting attack

**Comment by EvgeniiMekhanik** [comment]
I think that rate limits for control frames still actual. It is faster to block clients if they exceeded control frame rate limits.

**Comment by krizhanovsky** [comment]
With this change we can probably protect against [CDN amplification attack](https://www.usenix.org/conference/usenixsecurity24/presentation/lin-ziyu). For now we have issue in webshield https://github.com/tempesta-tech/webshield/issues/11 , but the logic seems heavyweight and it'd be better to protect on Tempesta FW side.

**Comment by EvgeniiMekhanik** [comment]
1.  common logic:
- implenent structure for collection trainging statistic
```
struct stats {
	struct percpu_counter num;
	struct percpu_counter sum;
	struct percpu_counter sumsq;
	s64 mean;
	s64 std;
};
```
- on each event during training period cal to update `sum` and `sumsq`
```
static inline void
tfw_training_mode_adjust_new_el(struct stats *s, u64 delta1, u64 delta2,
				bool new_client)
{
	if (new_client)
		percpu_counter_add(&s->num, 1);
	percpu_counter_add(&s->sum, delta1);
	percpu_counter_add(&s->sumsq, delta2);
}
```
- during switching from training to defence period calculate mean and std
```

static inline void
__calculate_mean_and_std(struct stats *s)
{
	s64 variance;
	u64 total_sum = 0;
	u64 total_sumsq = 0;
	u32 num_clients = 0;

	num_clients = percpu_counter_sum(&s->num);
	total_sum = percpu_counter_sum(&s->sum);
	total_sumsq = percpu_counter_sum(&s->sumsq);

	if (unlikely(!num_clients))
		return;

	s->mean = (total_sum << SCALE_SHIFT) / num_clients;
	variance = ((total_sumsq << SCALE_SHIFT) / num_clients) -
		((s->mean * s->mean) >> SCALE_SHIFT);
	s->std = int_sqrt64(variance << SCALE_SHIFT);
}
```
- during defence period on each event cal
```
static inline bool
__calculate_z_score(u64 val, struct stats *s, s64 *z_score)
{
	if (unlikely(!s->std))
		return false;

	*z_score = ((s64)(val << SCALE_SHIFT) - s->mean) / s->std;
	return true;

}
```
- compare calculated z_score from config value and block client if current z_score is greater then config value.

2. For new client connections we call all this functions under ra->lock, so we don;t need any atomics.
    add `conn_max` field to the TfwClient structure. In frang during incrementing conn_cur counter update conn_max and
    if became greater than previous one call tfw_training_mode_adjust_new_el
```
       old_max = cli->conn_max;
	if (conn_curr <= old_max)
		return true;
	cli->conn_max = conn_curr;
	delta1 = conn_curr - old_max;
	delta2 = (u64)conn_curr * conn_curr - (u64)old_max * old_max;
	tfw_training_mode_adjust_conn_num(delta1, delta2, new_client);
```

3. For cpu calculation imlement
```
typedef struct {
	TfwCpuHistoryBucket buckets[CPU_HISTORY_WINDOW];
	u64 ts;
} TfwCpuHistory;
``` to calculate cpu usage pre cliding window.

Calculate begin_time at the beginning of `tfw_tls_connection_recv` for tls connections and  `tfw_http_req_process` for plain http. Calculate end_time at the end of this functions. Pass time difference as cpu usage for z_score calculation. Use total cpu use per window as a value to calculate z-score.

**Comment by EvgeniiMekhanik** [comment]
**Training Mode:**
Collects per-client metrics on each new event aggregates global statistics: number of samples (num of clients),  sum of values (sum), sum of squares (sumsq). At the end of training computes mean and standard deviation, which will be used for z-score calculation in defence mode.
**Defence Mode**
Each new observation is evaluated using z=std / (x−mean). If z > configured_threshold he event is considered anomalous. 
Reject request / connection, drop connection with TCP RST and optionally block client by IP.
**Disabled Mode**
Internal state used during transitions. Ensures safe updates of shared data (via RCU synchronization). May be it's better to implement this state also, not only as internal state, to prevent any additional calculations, when it is not necessary.

**Connection Count Tracking**
In`TfwClient` structure we additionally store `unsigned int conn_max`, `int conn_curr` and `unsigned int conn_training_epoch`. We don't need any lock here, because all this fields updated under private `ra->lock` in frang.
We use new implemented function `tfw_client_training_adjust_conn_num` both for training and  defence mode.
**Training mode**
`conn_curr` is incremented/decremented.
Track maximum concurrent connections (`conn_max`). When max increases - compute `delta1 = new_max - old_max` and `delta2 = new_max² - old_max²` and use this values to update `sum` and `sumsq`.
**Defence mode**
Track `conn_curr` on each new opened connection. Calculate `z = (conn_curr - mean) / std` if `z > threshold` reject connection and block client by IP if necessary.
**Epoch handling**
Each connection tagged with training_epoch to prevents mixing old and new training data (we add new field to `tempesta_sock` and save epoch in this field). When connection closing we don't update `conn_curr` in case when connection belongs to previous epoch. (When connection is opening it always belongs to new epoch if trainging enabled!).
**Current method and alternatives**
In current approach during trainging mode we track maximum concurrent connections per client, in defence mode we compare current connections count (`conn_curr`) against a distribution of per-client maximum.
✔ effective for burst detection
❌ max vs current mismatch
❌ sensitivity to outliers (one client opens 10000 connections other clients 1 connection during trainging.)
❌ Not good against syn flood. If client open/close connections very fast, such client will not be blocked, because current connection count will be low.
**Alternatives:**
1. Use distribution of current concurrent connections - in trainging mode every `conn_curr` update stats, in defence mode compare `conn_curr` against this distribution. 
✔ Consistent model (same metric
❌ Noisy signal
❌ Dominated by low values (many idle clients, if they close there connections durig training)
2. Z-score on max, check only max
✔ Fully consistent model
✔ Low false positives
❌ Slow reaction
❌ Attack can stay just below max

In all this cases we don't take into account time awareness (100 connections for 1 second, 100 connections for 1 hour).
May be it will be good to calculate z-score for connection rate also (or we rely on frang for this case?).

**Request Count Tracking (Non-idempotent)**
We implement `TfwTrainingStat` structure to track all trainging events except connections.
```
/*
 * max		- maximum observed value of the tracked metric within the
 *		  current training epoch (e.g. peak number of in-flight
 *		  non-idempotent requests);
 * curr		- current value of the tracked metric;
 * lock		- spinlock for serialized reset of @max and @curr when a
 *		  new training epoch starts.
 * @epoch	- training epoch identifier. Compared against the global
 *		  @g_training_epoch to detect epoch change and trigger
 *		  reinitialization of @max and @curr.
 */
typedef struct {
	atomic64_t	max;
	atomic64_t	curr;
	spinlock_t	lock;
	unsigned int	epoch;
} TfwTrainingStat;
```
We use new implemented function `tfw_client_training_adjust_req_num` both for training and  defence mode.
**Training mode**
Track `curr` - current in-flight non-idempotent requests. Increment `curr` in `tfw_http_req_enlist`, decrement in  `tfw_http_req_nip_delist`. Also track `max` maximum count  in-flight non-idempotent requests per client. When max increases update global trainging stats, same as we do it for connections (`delta1 = new_max - old_max` and `delta2 = new_max² - old_max²`).
**Defence mode**
Change signature for `tfw_http_req_enlist` from `void` to `bool`.  Call `tfw_client_training_adjust_req_num` on each new non-idempotent request, calculate z-score, return false if `z > threshold`. `tfw_http_req_enlist` is called from `tfw_http_req_fwd` and `tfw_http_req_fwd_resched`, this functions now return T_BLOCK if `tfw_http_req_enlist` fails.
Callers of `tfw_http_req_fwd` and `tfw_http_req_fwd_resched` send 403 error response, drop client connection with TCP RST and block client by IP if these functions return T_BLOCK.
**Epoch handling**
Each request tagged with `training_epoch` to prevent mixing old and new training data (we add new field to `request` structure and save epoch in this field). When request removed from server connection queue we don't update `curr` field in case when request belongs to previous epoch. (When request added to server connection queue it always belongs to new epoch if trainging enabled!).
**Current method and alternatives**
The same problems and altgernatives as for connections.

**CPU Tracking**
In addition to `TfwTrainingStat` implement structure and per-cpu array of this structures.
```
/**
 * Exponential moving average (EMA) tracker for per-CPU time usage.
 *
 * The structure is used to accumulate execution time deltas and maintain
 * a smoothed estimate (EMA) of CPU consumption.
 *
 * @last_ts	- timestamp of the last update (in ns). Used to compute
 *		  time delta between consecutive measurements;
 * @ema		- current exponential moving average of CPU usage;
 * @pending_cpu	- accumulated raw CPU time (in ns) since the last EMA
 *		  update. This value is periodically folded into @ema;
 */
typedef struct {
    u64 last_ts;
    s64 ema;
    u64 pending_cpu;
} TfwCpuEma;
```
Save time at the beginning of SoftIRQ shot and  check CPU usage at the end of SoftIRQ shot (to prevent perfomance regression in case when we do it on each request) .
**Training mode**
Calculate `delta_cpu = now - begin_time;`, update CPU ema.
```
/**
 * Update per-client CPU usage EMA.
 * @cpu_ema: per-CPU EMA state for the client.
 * @delta_cpu: CPU time consumed since the last measurement (in ns).
 *
 * Accumulates raw CPU time in @pending_cpu and periodically folds it
 * into the exponential moving average (@ema).
 *
 * The update is performed only if enough time (@min_time_to_adjust)
 * has passed since the previous update to avoid excessive noise and
 * high-frequency recalculations.
 *
 * The function:
 *   - computes elapsed time (@dt);
 *   - converts accumulated CPU time into normalized usage value;
 *   - applies time-based decay (older history loses weight);
 *   - updates EMA using a combination of decay and smoothing factor.
 */
static void
tfw_client_update_cpu_ema(TfwCpuEma *cpu_ema, u64 delta_cpu)
{
	u64 now = ktime_get_ns();
	u64 dt = now - cpu_ema->last_ts;
	u64 usage, decay, total_cpu = 0;
	static const u64 time_to_forget_ns = 100000000;
	static const u64 min_time_to_adjust = 1000;
	static const unsigned int ema_alpha_shift = 4;

	cpu_ema->pending_cpu += delta_cpu;
	if (unlikely(dt < min_time_to_adjust))
		return;

	cpu_ema->last_ts = now;
	swap(cpu_ema->pending_cpu, total_cpu);
	usage = (total_cpu << SCALE_SHIFT) / dt;
	decay = (dt << SCALE_SHIFT) / time_to_forget_ns;

	if (decay > (1 << SCALE_SHIFT))
		decay = 1 << SCALE_SHIFT;
	cpu_ema->ema = cpu_ema->ema *
		((1 << SCALE_SHIFT) - decay) >> SCALE_SHIFT;
	cpu_ema->ema += ((s64)usage - (s64)cpu_ema->ema) >> ema_alpha_shift;
}
```
Pass `delta = new_ema - prev_ema` to  `tfw_client_training_adjust_cpu_num` which do the same as ` `tfw_client_training_adjust_req_num`.
**Defence mode**
In defence mode use `delta_ema` on each SoftIRQ shot to calculate `z = (delta_ema - mean) / std` and if calculated `z > threshold` reject connection with TCP RST and block client by IP if necessary.
**Current method and alternatives**

**Alternatives**
1. Use raw CPU time
✔ simple
✔ accuracy
❌ very noisy
❌ strong peaks
❌ Bad normalization
2. Sliding window average (store CPU usage for the last N ms)
3. Use `ema` directly. Currently we measure change, not level (constant high CPU → delta ≈ 0 → no detection).
